runtime/io.c

```diff
@@ -672,21 +682,24 @@ CAMLprim value caml_ml_close_channel(value vchannel)
   /* For output channels, must have flushed before */
   struct channel * channel = Channel(vchannel);
 
-  /* Ensure that every read or write on the channel will cause an
-     immediate caml_flush_partial or caml_refill, thus raising a Sys_error
-     exception */
-  channel->curr = channel->max = channel->end;
+  Lock(channel);
 
   /* If already closed, we are done */
-  if (channel->fd == -1) return Val_unit;
+  if (channel->fd != -1) {
+    fd = channel->fd;
+    channel->fd = -1;
+    caml_enter_blocking_section_no_pending();
+    result = close(fd);
+    caml_leave_blocking_section();
 
-  fd = channel->fd;
-  channel->fd = -1;
-  caml_enter_blocking_section_no_pending();
-  result = close(fd);
-  caml_leave_blocking_section();
+    if (result == -1) caml_sys_error (NO_ARG);
 
-  if (result == -1) caml_sys_error (NO_ARG);
+    /* Resetting curr makes it clear that this is a closed channel without
+     * unflushed data, so that no warning will be emitted by the channel's
+     * finalizer. */
```

Comment:
The comment is inexact? We may still have unflushed data on the buffer, since `caml_ml_close_channel` does not flush the buffer.

Reply:
It is a precondition that output channels should be flushed before calling `caml_ml_close_channel` (as says the comment at the beginning of the function), and `stdlib.ml` abides by this as the only way to close an `out_channel` is `close_out` which first flushes it.

Reply:
However, I have removed this comment in light of your remarks above.